src/neon_token.rs

```diff
@@ -107,3 +100,152 @@ async fn check_token_account(id: &ReqId) -> Result<()> {
 
     Ok(())
 }
+
+#[derive(Debug)]
+pub struct AirdropCapExceeded {
+    requested: u64,
+    limit: u64,
+}
+
+impl std::fmt::Display for AirdropCapExceeded {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Requested value {} exceeds the limit {}", self.requested, self.limit)
+    }
+}
+
+#[derive(Debug)]
+pub enum AirdropLimiterError {
+    BadRequest,
+    CapExceeded(AirdropCapExceeded),
+    BadConversion,
+}
+
+impl std::error::Error for AirdropLimiterError {}
+
+impl std::fmt::Display for AirdropLimiterError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::BadRequest => write!(f, "Bad airdrop request"),
+            Self::CapExceeded(err) => write!(f, "{}", err),
+            Self::BadConversion => write!(f, "Failed to convert to fractional token value"),
+        }
+    }
+}
+
+pub struct AirdropLimiter {
+    trusted_proxies: HashSet<IpAddr>,
+    blacklist: HashSet<IpAddr>,
+    // Token amounts in fractional units.
+    ip_cache: HashMap<IpAddr, u64>,
+    time_slice: Duration,
+    per_request_cap: u64,
+    per_time_cap: u64,
+}
+
+impl AirdropLimiter {
+    pub fn new(
+        trusted_proxies: HashSet<IpAddr>,
+        blacklist: HashSet<IpAddr>,
+        time_slice: Duration,
+        per_request_cap: u64,
+        per_time_cap: u64,
+    ) -> Self {
+        Self {
+            trusted_proxies,
+            blacklist,
+            ip_cache: Default::default(),
+            time_slice,
+            per_request_cap,
+            per_time_cap,
+        }
+    }
+
+    pub fn time_slice(&self) -> Duration {
+        self.time_slice
+    }
+
+    pub fn clear_cache(&mut self) {
+        self.ip_cache.clear();
+    }
+
+    pub fn check_cache(
+        &mut self,
+        req: &HttpRequest,
+        airdrop: &Airdrop
+    ) -> Result<(), AirdropLimiterError> {
+        let peer = self.get_peer(req)?;
+        let request_amount = Self::parse_amount(airdrop)?;
+        if request_amount > self.per_request_cap {
+            error!("Airdrop request capped at {}", self.per_request_cap);
+            return Err(AirdropLimiterError::CapExceeded(AirdropCapExceeded {
+                requested: request_amount,
+                limit: self.per_request_cap,
+            }));
+        }
+        let total = self.ip_cache
+            .entry(peer)
+            .and_modify(|total| *total = total.saturating_add(request_amount))
+            .or_insert(request_amount);
```

Comment:
This seems a little cleaner:
```suggestion
        let total = self.ip_cache.entry(peer).or_default();
        *total = total.saturating_add(request_amount);
```

Reply:
sure